src/renderer_common/render_markers_symbolizer.cpp

```diff
@@ -130,18 +129,69 @@ struct render_marker_symbolizer_visitor
-        auto const& r_attributes = get_marker_attributes(*stock_vector_marker, s_attributes);
+        std::shared_ptr<svg_attribute_type> r_attributes;
+
+        // Look up the feature/symbolizer attributes from the cache.
+        // We are using raw symbolizer pointer as a cache key. This is a hack but
+        // works in practice due to symbolizers being created all at once during style loading.
+        markers_symbolizer const* attr_key = &sym_;
+
+        auto attr_it = cached_attributes_.find(attr_key);
+        if (attr_it == cached_attributes_.end())
+        {
+            svg_attribute_type s_attributes;
+            r_attributes = std::make_shared<svg_attribute_type>(get_marker_attributes(*stock_vector_marker, s_attributes));
+
+            // We can only cache the attributes using the given key if no expressions are used in properties.
+            // Otherwise the expressions may refer to feature-specific values.
+            bool cacheable = std::all_of(
+                sym_.properties.begin(), sym_.properties.end(),
+                [](symbolizer_base::cont_type::value_type const& key_prop) { return !is_expression(key_prop.second); }
+            );
+            if (cacheable)
+            {
+                if (cached_attributes_.size() > cache_size)
+                {
+                    cached_attributes_.erase(cached_attributes_.begin());
+                }
+                attr_it = cached_attributes_.emplace(attr_key, r_attributes).first;
+            }
+        }
+        else
+        {
+            r_attributes = attr_it->second;
+        }
 
         // special case for simple ellipse markers
         // to allow for full control over rx/ry dimensions
         if (filename_ == "shape://ellipse"
            && (has_key(sym_,keys::width) || has_key(sym_,keys::height)))
         {
-            marker_ptr = std::make_shared<svg_storage_type>();
-            is_ellipse = true;
+            // Ellipses are built procedurally. We do caching of the built ellipses, this is useful for rendering stages
+            std::tuple<double, double, double> marker_key(
+                get<double>(sym_, keys::width, feature_, common_.vars_, -std::numeric_limits<double>::infinity()),
+                get<double>(sym_, keys::height, feature_, common_.vars_, -std::numeric_limits<double>::infinity()),
+                get<double>(sym_, keys::stroke_width, feature_, common_.vars_, -std::numeric_limits<double>::infinity())
+            );
+
+            auto marker_it = cached_ellipses_.find(marker_key);
+            if (marker_it == cached_ellipses_.end())
+            {
+                marker_ptr = std::make_shared<svg_storage_type>();
+
+                vertex_stl_adapter<svg_path_storage> stl_storage(marker_ptr->source());
+                svg_path_adapter svg_path(stl_storage);
+                build_ellipse(sym_, feature_, common_.vars_, *marker_ptr, svg_path);
+
+                if (cached_ellipses_.size() > cache_size)
```

Comment:
Seems this "breaks" the condition imposed by cache_size. I'd expect to be 128 for both elipses and attributes together, not each.